### PR TITLE
[Feat] Layerwise casting hook

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Introduces a layerwise casting method that lets users run a model in separate storage and compute dtypes, leading to memory savings. See internal thread:
https://huggingface.slack.com/archives/C01Q6JPP6NA/p1739198698861009

I tested a sufficiently large dummy model with it and here are some numbers: 

| Batch Size | Layerwise Casting | Autocast | Model Memory Footprint (MB) | Peak Inference Memory Allocated (GB) |
|:---------:|:----------------:|:--------:|:---------------------------:|:------------------------------------:|
|     1     |       True       |   False  |           428.23           |              1635.86               |
|     1     |       False      |   True   |          1633.56           |              2475.92               |
|     4     |       True       |   False  |           408.39           |              1639.11               |
|     4     |       False      |   True   |          1633.56           |              2539.73               |
|    16     |       True       |   False  |           408.39           |              1649.02               |
|    16     |       False      |   True   |          1633.56           |              2848.93               |

<details>
<summary>script</summary>

```py
import torch
from transformers import SiglipVisionModel
from accelerate.big_modeling import apply_layerwise_casting
from contextlib import nullcontext
import argparse
import json

SUPPORTED_PYTORCH_LAYERS = (
    torch.nn.Conv1d, torch.nn.Conv2d, torch.nn.Conv3d,
    torch.nn.ConvTranspose1d, torch.nn.ConvTranspose2d, torch.nn.ConvTranspose3d,
    torch.nn.Linear,
)

def get_memory_footprint(model, return_buffers=True):
    mem = sum([param.nelement() * param.element_size() for param in model.parameters()])
    if return_buffers:
        mem_bufs = sum([buf.nelement() * buf.element_size() for buf in model.buffers()])
        mem = mem + mem_bufs
    return mem


def get_dummy_inputs(batch_size=1, target_dtype=torch.float32):
    inputs = {"pixel_values": torch.randn(batch_size, 3, 384, 384).to("cuda", target_dtype)}
    return inputs

def check_linear_dtype(module, storage_dtype, compute_dtype):
    for name, submodule in module.named_modules():
        if not isinstance(submodule, SUPPORTED_PYTORCH_LAYERS):
            continue
        dtype_to_check = storage_dtype
        if getattr(submodule, "weight", None) is not None:
            assert submodule.weight.dtype == dtype_to_check, submodule.weight.dtype
        if getattr(submodule, "bias", None) is not None:
            assert submodule.bias.dtype == dtype_to_check, submodule.bias.dtype


@torch.no_grad()
@torch.inference_mode()
def run(model, inputs):
    model(**inputs)
    model_mem_footprint = get_memory_footprint(model) / 1024**2
    peak_inference_memory_allocated_mb = torch.cuda.max_memory_allocated() / 1024**2
    return {
        "model_mem_footprint_mb": model_mem_footprint,
        "peak_inference_memory_allocated_gb": peak_inference_memory_allocated_mb
    }


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--batch_size", type=int, default=1)
    parser.add_argument("--apply_layerwise_casting", action="store_true")
    parser.add_argument("--use_autocast", action="store_true")
    args = parser.parse_args()

    model = SiglipVisionModel.from_pretrained("google/siglip-so400m-patch14-384").to("cuda")
    
    if args.apply_layerwise_casting:
        inputs = get_dummy_inputs(batch_size=args.batch_size, target_dtype=torch.bfloat16)
        apply_layerwise_casting(model, storage_dtype=torch.float8_e5m2, compute_dtype=torch.bfloat16)
        context = nullcontext()
    elif args.use_autocast:
        inputs = get_dummy_inputs(batch_size=args.batch_size)
        context = torch.autocast("cuda", dtype=torch.bfloat16, enabled=True)
    
    with context:
        output = run(model, inputs)

    print(f"{args.batch_size=}, {args.apply_layerwise_casting=}, {args.use_autocast=}")
    print(json.dumps(output, indent=4))
```

</details>

I haven't updated the tests yet as this PR is for gathering feedback to know if it's headed in the right direction. 

## TODOs

- [x] Incorporate `skip_modules_pattern` arg
- [ ] Docs
- [x] Tests

### CODE REVIEW

File: src/accelerate/big_modeling.py [hunk 1]
```diff
@@ -638,3 +639,43 @@ def load_checkpoint_and_dispatch(
...
+        skip_modules_classes (`Tuple[Type[torch.nn.Module], ...]`, defaults to `None`):
+            A list of module classes to skip during the layerwise casting process.
+        non_blocking (`bool`, defaults to `False`):
+            If `True`, the weight casting operations are non-blocking.
+    """
+    add_hook_to_module(
+        module,
+        LayerwiseCastingHook(storage_dtype=storage_dtype, compute_dtype=compute_dtype, non_blocking=non_blocking),
+        append=True,
+    )
```
Comment (SunMarc) [inline src/accelerate/big_modeling.py:681]: for the submodules, to imitate the other hooks, maybe create a function starting with "attach" maybe `attach_layerwise_casting_hooks` ?

**Review by SunMarc — COMMENTED** [review]
So far so good, thanks for stating this work !

**Review by SunMarc — COMMENTED** [review]
Thanks for the work @sayakpaul , just a small edge case

File: tests/test_hooks.py [hunk 2]
```diff
@@ -400,3 +416,27 @@ def test_add_remove_hook_fx_graph_module(self):
...
+            test_model,
+            storage_dtype=storage_dtype,
+            compute_dtype=compute_dtype,
+            skip_modules_pattern=skip_modules_pattern,
+        )
+        patterns_to_check = skip_modules_pattern if skip_modules_pattern else None
+        self.check_dtype_for_layerwise_upcasting(test_model, storage_dtype, compute_dtype, patterns_to_check)
+
+        with torch.no_grad():
+            _ = test_model(inputs)
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply tests/test_hooks.py:442]: You mean we call 

```py
attach_layerwise_casting_hooks(
    test_model,
    storage_dtype=storage_dtype,
    compute_dtype=compute_dtype,
    skip_modules_pattern=None,
)
```

and the `test_model` isn't downcasted at all? Not sure how we could do that. Could you elaborate more?

**Comment by sayakpaul** [comment]
@SunMarc seems like there's something wrong with the CI:
https://github.com/huggingface/accelerate/actions/runs/14529383877/job/40766557692?pr=3427#step:3:11

**Comment by SunMarc** [comment]
> @SunMarc seems like there's something wrong with the CI:
> https://github.com/huggingface/accelerate/actions/runs/14529383877/job/40766557692?pr=3427#step:3:11

Yeah, should be fixed in this PR ! Thanks for the heads-up.

File: tests/test_hooks.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply tests/test_hooks.py:442]: We don't have any default modules to not skip. I don't know if we can do what you're asking for without defining a default module name that will never get downcasted.

[the remark this answers is not in the document]
Comment (SunMarc) [reply tests/test_hooks.py:442]: You have that no ? SUPPORTED_PYTORCH_LAYERS_FOR_UPCASTING

[the remark this answers is not in the document]
Comment (sayakpaul) [reply tests/test_hooks.py:442]: Ah okay. Sorry for the confusion.

[the remark this answers is not in the document]
Comment (sayakpaul) [reply tests/test_hooks.py:458]: Done

File: src/accelerate/big_modeling.py [hunk 3]
```diff
@@ -673,34 +672,51 @@ def attach_layerwise_casting_hooks(
...
-            append=True,
-        )
-        return
-    else:
-        _attach_layerwise_casting_hooks(
-            module, storage_dtype, compute_dtype, skip_modules_pattern, skip_modules_classes, non_blocking
-        )
+    _attach_layerwise_casting_hooks(
+        module, storage_dtype, compute_dtype, skip_modules_pattern, skip_modules_classes, non_blocking
+    )
```
Comment (SunMarc) [inline src/accelerate/big_modeling.py:711]: if you think that in the future, we will modify `attach_layerwise_casting_hooks` without touching _attach_layerwise_casting_hooks, then we can keep it this way. Otherwise, let's just put everything in `attach_layerwise_casting_hooks` wdyt ?

File: tests/test_hooks.py [hunk 4]
```diff
@@ -60,17 +60,33 @@ def post_forward(self, module, output):
...
+                        self.assertNotEqual(tensor, storage_dtype)
+                continue
+
+            if patterns_to_check and any(re.search(pat, name) for pat in patterns_to_check):
+                expected = compute_dtype
+            else:
+                expected = storage_dtype
+
+            for _, tensor in attrs:
+                self.assertEqual(tensor.dtype, expected)
```
Comment (SunMarc) [inline tests/test_hooks.py:89]: Not quite there yet : 

- if not isinstance(submodule, SUPPORTED_PYTORCH_LAYERS_FOR_UPCASTING) -> The weights should be in loading_dtype instead of storage_dtype in reality. 
- if there is a pattern, this is the same as the first case so loading_dtype instead of `compute_dtype`
- else: -> the module should indeed be in storage_dtype

The modules here are never in `compute_dtype` as this would require to check the module dtype just before the forward of this particular module. 

LMK if this is not clear !

**Review by SunMarc — COMMENTED** [review]
A few nits !

File: src/accelerate/big_modeling.py (hunk 3 above)
Comment (sayakpaul) [reply src/accelerate/big_modeling.py:711]: I think as is currently is okay.

File: tests/test_hooks.py (hunk 4 above)
Comment (sayakpaul) [reply tests/test_hooks.py]: How about now
